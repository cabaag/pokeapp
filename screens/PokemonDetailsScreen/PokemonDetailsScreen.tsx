import { RouteProp, useNavigation } from '@react-navigation/native';
import { Card, Col, Container, Content, Row, Text, Header, Button, Left, Icon, Body, Spinner } from 'native-base';
import * as React from 'react';
import { Image, StyleSheet } from 'react-native';
import { Pokemon } from '../../types/Pokemon';
import { RootStackParamList } from '../../types/Stacks';

type PokemonDetaulsScreenRouteProp = RouteProp<RootStackParamList, 'PokemonDetailsScreen'>;

const styles = StyleSheet.create({
  logo: {
    height: 60,
    width: 140,
    flex: 1,
    alignItems: 'center',
  },
  sprite: {
    width: 200,
    height: 200
  },
  body: {
    flex: 1,
    alignContent: 'center',
    justifyContent: 'center',
    alignItems: 'center'
  }
})


export interface PokemonDetailsScreenProps {
  pokemon: Pokemon;
  route: PokemonDetaulsScreenRouteProp;
}

const PokemonDetailsScreen: React.FC<PokemonDetailsScreenProps> =
  ({ route }: PokemonDetailsScreenProps) => {
    const navigation = useNavigation();

    const { pokemon } = route.params;
    return (
      <Container>
        <Header>
          <Left>
            <Button onPress={() => navigation.goBack()} transparent>
              <Icon name='arrow-back' />
            </Button>
          </Left>
          <Body style={{  }}>
            <Image source={require('../../assets/images/pokemon.png')} style={styles.logo} />
          </Body>
        </Header>
        <Content padder>
          <Card>
            <Body>
              <Row>
                <Col>
                  {
                    pokemon?.sprites?.front_default ?
                      <Image source={{ uri: pokemon.sprites.front_default }} style={styles.sprite} />
                      : <Spinner />
                  }
                </Col>
                <Col style={{alignContent: 'flex-start', alignItems: 'flex-start'}}>
                  <Row>
                    <Text>
                      #
                      {pokemon.id}
                    </Text>
                  </Row>
                  <Row>
                    <Text style={{ fontWeight: 'bold', textTransform: 'capitalize' }}>
                      {pokemon?.name}
                    </Text>
                  </Row>
                  <Row>
                    <Text>
                      Height:
                      {pokemon.height / 10}
                      m
                    </Text>
                  </Row>
                  <Row>
                    <Text>
                      Weight:
                      {pokemon.weight / 10}
                      kg.
                    </Text>
                  </Row>
                </Col>
              </Row>
            </Body>
          </Card>
        </Content>
      </Container>
    );
  }

export default PokemonDetailsScreen;