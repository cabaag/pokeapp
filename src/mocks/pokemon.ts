import { Pokemon } from '../types/Pokemon';

const PokemonMock: Pokemon = {
  base_experience: 0,
  height: 10,
  weight: 10,
  name: 'test',
  sprites: {
    front_default: '',
    front_shiny: '',
  },
  stats: [
    {
      base_stat: 10,
      effort: 0,
      stat: {
        name: 'stat test',
        url: '',
      },
    },
  ],
  id: 0,
};

export { PokemonMock };
