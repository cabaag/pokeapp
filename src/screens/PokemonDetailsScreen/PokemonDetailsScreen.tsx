import { RouteProp, useNavigation } from '@react-navigation/native';
import { Body, Card, Col, Container, Content, Row, Spinner, Text } from 'native-base';
import React, { useEffect, useLayoutEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Image, StyleSheet, View } from 'react-native';
import Slider from '@react-native-community/slider';
import typesMapper from '../../utils/typesMapper';
import { RootStackParamList } from '../../types/Stacks';
import { retrievePokemon } from '../../services/Pokemon';
import { Pokemon } from '../../types/Pokemon';

type PokemonDetaulsScreenRouteProp = RouteProp<RootStackParamList, 'PokemonDetailsScreen'>;

const styles = StyleSheet.create({
  sprite: {
    width: 150,
    height: 150
  },
  line: {
    backgroundColor: '#ddd',
    height: 1,
    flex: 1,
    borderRadius: 10,

  },
  badge: {
    borderRadius: 50,
    width: 10,
    height: 10,
    margin: 10,
  }

})

export interface PokemonDetailsScreenProps {
  route: PokemonDetaulsScreenRouteProp;
}

const PokemonDetailsScreen: React.FC<PokemonDetailsScreenProps> =
  ({ route }: PokemonDetailsScreenProps) => {
    const navigation = useNavigation();
    const { t } = useTranslation();
    const { pokemonName } = route.params;
    const [loading, setLoading] = useState(true)
    const [pokemon, setPokemon] = useState<Pokemon>();

    useEffect(() => {
      retrievePokemon(pokemonName).then(setPokemon).finally(() => {
        setLoading(false)
      })
    }, [])

    useLayoutEffect(() => {
      navigation.setOptions({
        title: pokemon?.name
      })
    }, [navigation, pokemon]);


    return (
      <Container>
        <Content padder>
          <Card>
            <Body style={{ alignItems: 'flex-start', paddingRight: 20, paddingLeft: 20 }}>
              <Row>
                <Col>
                  {loading && <Spinner />}
                  {
                    !loading && pokemon?.sprites?.front_default ?
                      <Image source={{ uri: pokemon.sprites.front_default }} style={styles.sprite} />
                      : null
                  }
                </Col>
                <Col style={{ alignContent: 'flex-start', alignItems: 'flex-start', paddingTop: 20 }}>
                  <Text>
                    #
                    {pokemon?.id}
                  </Text>
                  <Text style={{ fontWeight: 'bold', textTransform: 'capitalize' }}>
                    {pokemon?.name}
                  </Text>
                  <Text>
                    {t('height')}
                    :
                    {' '}
                    {(pokemon?.height ?? 1) / 10}
                    m
                  </Text>
                  <Text>
                    {t('weight')}
                    :
                    {' '}
                    {(pokemon?.weight ?? 1) / 10}
                    kg
                  </Text>
                  <Row>
                    {pokemon?.types?.map(type => (
                      <View
                        key={type.slot}
                        style={{
                          ...styles.badge,
                          backgroundColor: typesMapper(type.type.name) as string
                        }}
                      />
                    ))}
                  </Row>
                </Col>
              </Row>
              <Row>
                <Row style={{ flexDirection: 'row', alignItems: 'center', alignContent: 'center' }}>
                  <Col style={styles.line} />
                  <Text style={{ fontSize: 20 }}>
                    {t('stadistics')}
                  </Text>
                  <Col style={styles.line} />
                </Row>
              </Row>
              <Col style={{ paddingTop: 10, paddingBottom: 10 }}>
                {
                  pokemon?.stats.map(stat => (
                    <Col key={stat.stat.name}>
                      <Row style={{ alignItems: 'center' }}>
                        <Text style={{ fontWeight: 'bold', width: 130 }}>
                          {t(stat.stat.name)}
                          {' '}
                          :
                          {stat?.base_stat}
                        </Text>
                        <Slider
                          maximumTrackTintColor="#3F51B5"
                          maximumValue={140}
                          minimumTrackTintColor="#3F51B5"
                          minimumValue={0}
                          style={{ width: 230, height: 40 }}
                          thumbTintColor='#3F51B5'
                          value={stat?.base_stat}

                        />
                      </Row>
                    </Col>
                  ))
                }
              </Col>
            </Body>
          </Card>
        </Content>
      </Container>
    );
  }

export default React.memo(PokemonDetailsScreen);