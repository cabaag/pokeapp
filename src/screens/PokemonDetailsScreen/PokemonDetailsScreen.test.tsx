import { shallow, ShallowWrapper } from 'enzyme';
import { PokemonMock } from 'mocks/pokemon';
import { Spinner } from 'native-base';
import React from 'react';
import PokemonDetailsScreen, { PokemonDetailsScreenProps } from './PokemonDetailsScreen';

const goBack = jest.fn();

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    goBack
  })
}));

jest.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (key: string) => key,
  }),
}));

const defaultProps: PokemonDetailsScreenProps = {
  route: {
    key: '',
    name: 'PokemonDetailsScreen',
    params: {
      pokemon: PokemonMock
    }
  },
}

function setup(customProps?: any): ShallowWrapper {
  return shallow(
    <PokemonDetailsScreen {...defaultProps} {...customProps} />
  )
}

describe('<PokemonDetailsScreen />', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = setup();
  })

  it('matchs snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should go back', () => {
    const backButton = wrapper.find("[data-test='goBack']")
    backButton.simulate('press');
    expect(goBack).toHaveBeenCalled();
  })

  it('should render spinner', () => {
    wrapper = setup({
      route: {
        key: '',
        name: 'PokemonDetailsScreen',
        params: {
          pokemon: null
        }
      }
    })
    expect(wrapper.find(Spinner)).toHaveLength(1);
  })

});