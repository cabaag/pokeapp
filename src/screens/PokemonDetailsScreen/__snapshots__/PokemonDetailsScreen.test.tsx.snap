// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonDetailsScreen /> matchs snapshot 1`] = `
<Styled(Container)>
  <Styled(Content)
    padder={true}
  >
    <Styled(Card)>
      <Styled(Body)
        style={
          Object {
            "alignItems": "flex-start",
            "paddingLeft": 20,
            "paddingRight": 20,
          }
        }
      >
        <RowNB>
          <ColumnNB>
            <Styled(Spinner) />
          </ColumnNB>
          <ColumnNB
            style={
              Object {
                "alignContent": "flex-start",
                "alignItems": "flex-start",
                "paddingTop": 20,
              }
            }
          >
            <Styled(Text)>
              #
              0
            </Styled(Text)>
            <Styled(Text)
              style={
                Object {
                  "fontWeight": "bold",
                  "textTransform": "capitalize",
                }
              }
            >
              test
            </Styled(Text)>
            <Styled(Text)>
              height
              :
               
              1
              m
            </Styled(Text)>
            <Styled(Text)>
              weight
              :
               
              1
              kg
            </Styled(Text)>
            <RowNB>
              <View
                key="1"
                style={
                  Object {
                    "backgroundColor": "#AB1F23",
                    "borderRadius": 50,
                    "height": 10,
                    "margin": 10,
                    "width": 10,
                  }
                }
              />
            </RowNB>
          </ColumnNB>
        </RowNB>
        <RowNB>
          <RowNB
            style={
              Object {
                "alignContent": "center",
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <ColumnNB
              style={
                Object {
                  "backgroundColor": "#ddd",
                  "borderRadius": 10,
                  "flex": 1,
                  "height": 1,
                }
              }
            />
            <Styled(Text)
              style={
                Object {
                  "fontSize": 20,
                }
              }
            >
              stadistics
            </Styled(Text)>
            <ColumnNB
              style={
                Object {
                  "backgroundColor": "#ddd",
                  "borderRadius": 10,
                  "flex": 1,
                  "height": 1,
                }
              }
            />
          </RowNB>
        </RowNB>
        <ColumnNB
          style={
            Object {
              "paddingBottom": 10,
              "paddingTop": 10,
            }
          }
        >
          <ColumnNB
            key="stat test"
          >
            <RowNB
              style={
                Object {
                  "alignItems": "center",
                }
              }
            >
              <Styled(Text)
                style={
                  Object {
                    "fontWeight": "bold",
                    "width": 130,
                  }
                }
              >
                stat test
                 
                :
                10
              </Styled(Text)>
              <ForwardRef(SliderComponent)
                disabled={false}
                inverted={false}
                maximumTrackTintColor="#3F51B5"
                maximumValue={140}
                minimumTrackTintColor="#3F51B5"
                minimumValue={0}
                step={0}
                style={
                  Object {
                    "height": 40,
                    "width": 230,
                  }
                }
                thumbTintColor="#3F51B5"
                value={10}
              />
            </RowNB>
          </ColumnNB>
        </ColumnNB>
      </Styled(Body)>
    </Styled(Card)>
  </Styled(Content)>
</Styled(Container)>
`;
